### ISSUE TITLE
Guest additions download failed after upgrading from 1.1.1 to 1.1.3

### ISSUE THREAD

**Issue by andrewsav-datacom** [hashicorp/packer#5713]
I have a packer template similar to this one: https://github.com/boxcutter/ubuntu

After upgrade to 1.1.3, I'm getting the following:

```
2017/12/18 17:14:03 ui: virtualbox-iso output will be in this color.
2017/12/18 17:14:03 ui: vmware-iso output will be in this color.
2017/12/18 17:14:03 ui: 
2017/12/18 17:14:03 Build debug mode: false
2017/12/18 17:14:03 Force build: false
2017/12/18 17:14:03 On error: 
2017/12/18 17:14:03 Preparing build: virtualbox-iso
2017/12/18 17:14:03 Preparing build: vmware-iso
2017/12/18 17:14:03 Starting build run: virtualbox-iso
2017/12/18 17:14:03 Running builder: virtualbox-iso
2017/12/18 17:14:03 [INFO] (telemetry) Starting builder virtualbox-iso
2017/12/18 17:14:03 Waiting on builds to complete...
2017/12/18 17:14:03 Starting build run: vmware-iso
2017/12/18 17:14:03 Running builder: vmware-iso
2017/12/18 17:14:03 [INFO] (telemetry) Starting builder vmware-iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 [DEBUG] builder/virtualbox: VBOX_MSI_INSTALL_PATH = C:\Program Files\Oracle\VirtualBox\
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 VBoxManage path: C:\Program Files\Oracle\VirtualBox\VBoxManage.exe
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Searching for file 'C:/Program Files (x86)/VMware/VMware Workstation/vmware.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Found file 'C:/Program Files (x86)/VMware/VMware Workstation/vmware.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Searching for file 'C:/Program Files (x86)/VMware/VMware Workstation/vmrun.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Found file 'C:/Program Files (x86)/VMware/VMware Workstation/vmrun.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Searching for file 'C:/Program Files (x86)/VMware/VMware Workstation/vmware-vdiskmanager.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Found file 'C:/Program Files (x86)/VMware/VMware Workstation/vmware-vdiskmanager.exe'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 VMware app path: C:/Program Files (x86)/VMware/VMware Workstation/vmware.exe
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 vmrun path: C:/Program Files (x86)/VMware/VMware Workstation/vmrun.exe
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 vdisk-manager path: C:/Program Files (x86)/VMware/VMware Workstation/vmware-vdiskmanager.exe
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Detected VMware WS version: 14
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Searching for file 'C:/Program Files (x86)/VMware/VMware Workstation/linux.iso'
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Found file 'C:/Program Files (x86)/VMware/VMware Workstation/linux.iso'
2017/12/18 17:14:03 ui: ==> vmware-iso: Downloading or copying ISO
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Acquiring lock to download: file://Volumes/Storage/software/ubuntu/ubuntu-16.04.3-server-amd64.iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Acquiring lock to download: http://releases.ubuntu.com/16.04.3/ubuntu-16.04.3-server-amd64.iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 VBoxManage --version output: 5.1.30r118389
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 VirtualBox version: 5.1.30
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Verifying checksum of E:\PackerCache\2b6eba7bfd46552a598360d9b28d3a584f91260d3fef906c8987bc74be115182.iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Found Default Guest Additions ISO: C:\Program Files\Oracle\VirtualBox/VBoxGuestAdditions.iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Guest additions URL: file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
2017/12/18 17:14:03 ui: ==> virtualbox-iso: Downloading or copying Guest additions
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Acquiring lock to download: file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
2017/12/18 17:14:03 ui:     virtualbox-iso: Downloading or copying: file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 Parsed URL: &url.URL{Scheme:"file", Opaque:"", User:(*url.Userinfo)(nil), Host:"C:", Path:"/Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso", RawPath:"", ForceQuery:false, RawQuery:"", Fragment:""}
2017/12/18 17:14:03 packer.exe: 2017/12/18 17:14:03 [DEBUG] Using local file: /Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
2017/12/18 17:14:03 ui:     virtualbox-iso: Error downloading: CreateFile /Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso: The system cannot find the path specified.
2017/12/18 17:14:03 ui error: ==> virtualbox-iso: Guest additions download failed.
2017/12/18 17:14:03 [INFO] (telemetry) ending virtualbox-iso
2017/12/18 17:14:03 [INFO] (telemetry) found error: Guest additions download failed.
2017/12/18 17:14:03 ui error: Build 'virtualbox-iso' errored: Guest additions download failed.
```

What could be the issue?

**Comment by SwampDragons** [comment]
I can't reproduce with 1.1.3 on windows.  Can you please share your exact config?

**Comment by andrewsav-datacom** [comment]
Attached: [ubuntu-1604.zip](https://github.com/hashicorp/packer/files/1569730/ubuntu-1604.zip)

@SwampDragons, thank you for looking at this.

**Comment by mwhooker** [comment]
possibly a dupe of #5700

**Comment by mwhooker** [comment]
and maybe not a dupe, but seems closely related to #5559

**Comment by andrewsav-datacom** [comment]
Not hyper-v here, but yes, looks very similar.

**Comment by mwhooker** [comment]
the builders use the same code under the hood

**Comment by andrewsav-datacom** [comment]
Works fine on VmWare builder though.

**Comment by mwhooker** [comment]
There was a fix in #5591 for this same error message, making me think these have a root cause in maybe #5573

**Comment by AndrewSav** [comment]
@SwampDragons can you reproduce it with the config provided?

**Comment by SwampDragons** [comment]
I haven't had a chance to try yet today; may not get to it until tomorrow but will try to get to it before Hashicorp's end of year break.

**Comment by SwampDragons** [comment]
I still can't repro.  I used ubuntu.json in your attached zip, since ubuntu1604.json wasn't a complete packer config.  Can I have your full debug logs?

**Comment by SwampDragons** [comment]
Dumb sanity-check question. Are you sure you're on 1.1.3 and not 1.1.2?

**Comment by andrewsav-datacom** [comment]
@SwampDragons definitely on 1.1.3. You cannot reproduce? I tried on two different machines and the results are the same.

**Comment by SwampDragons** [comment]
Yeah, still can't.  <img width="1332" alt="screen shot 2017-12-20 at 3 32 34 pm" src="https://user-images.githubusercontent.com/1008838/34233579-148dedce-e59b-11e7-84ba-9f20b5275fd4.png">

**Comment by andrewsav-datacom** [comment]
Okay, let me update VirtualBox to 5.2.2, although that should not matter. And I'll also re-download the packer executuable just to be sure, long shot though.

**Comment by andrewsav-datacom** [comment]
@SwampDragons 

```
2017/12/19 12:16:53 [INFO] Packer version: 1.1.3
2017/12/19 12:16:53 Packer Target OS/Arch: windows amd64
2017/12/19 12:16:53 Built with Go Version: go1.9
```
Why does your screenshot says go1.8.1 ?
We are clearly running different executables.

**Comment by SwampDragons** [comment]
I built it locally from our v1.1.3 release tags just because that's how I normally develop.  I can grab it from our release page and validate that this isn't an issue, though.

**Comment by andrewsav-datacom** [comment]
@SwampDragons thank you, that would be nice to double check.

**Comment by SwampDragons** [comment]
Just checked with the one from the downloads site; still no issues for me.

**Comment by SwampDragons** [comment]
so just looking at where the logs are in the code, it seems like you must be failing here: https://github.com/hashicorp/packer/blob/master/common/download.go#L134 which made me flag on the idea that maybe shell versions matter?  But I don't think os.Stat should cause windows any problems.

**Comment by andrewsav-datacom** [comment]
And I'm still failing with VirtualBox 5.2.2 and packer re-downloaded from the hashicorp site (used choco package before that).

**Comment by SwampDragons** [comment]
Not that I know how this will help at this point, but what version of windows are you on again?

**Comment by andrewsav-datacom** [comment]
Tested in Windows 2016 Server and Windows 10.


Win 10:
```
Platform ServicePack Version      VersionString
-------- ----------- -------      -------------
 Win32NT             10.0.16299.0 Microsoft Windows NT 10.0.16299.0
```

Win 2016:
```
Platform ServicePack Version      VersionString
-------- ----------- -------      -------------
 Win32NT             10.0.14393.0 Microsoft Windows NT 10.0.14393.0
```

**Comment by SwampDragons** [comment]
Shoot, me too. I'm at a total loss here.  At this point all that is coming through my mind is "permissions...?"

**Comment by andrewsav-datacom** [comment]
@SwampDragons can you try this.

Put the templates and the copy of packer.exe on a different disk. That is not on the disk where Virtual Box is installed, and try running it.

When I move the packer template to C: it starts working here as well.

**Comment by SwampDragons** [comment]
Oooooh interesting.

**Comment by andrewsav-datacom** [comment]
@SwampDragons so on https://github.com/hashicorp/packer/blob/master/common/download.go#L134 it stats `/Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso` and since it's on the wrong drive it cannot find it. Mystery solved, the only thing left is to make a fix ;)

**Comment by andrewsav-datacom** [comment]
**Summary**
The root of this issue is that net/url is used to parse file paths. It is not designed for this, and should not be used this way.

See here: https://github.com/golang/go/issues/13276#issuecomment-157074191

[This](https://github.com/hashicorp/packer/blob/master/common/config.go#L49) is the function that needs to be rewritten to address that.

**What's went wrong?**

Note this piece of code that was removed in a [recent commit](https://github.com/hashicorp/packer/commit/3a9dfb5b1897bc3fd795ca1e837bd7dd419221eb#diff-6c015c1d07296dbab7d1a69c15130511):

```
	// This is to work around issue #5927. This can safely be removed once
	// we distribute with a version of Go that fixes that bug.
	//
	// See: https://code.google.com/p/go/issues/detail?id=5927
	if url.Path != "" && url.Path[0] != '/' {
		url.Path = "/" + url.Path
	}
```

This piece of code was already [removed](https://github.com/hashicorp/packer/commit/44008c321b1df7192d5be1c6a6ea3c45c706a29a#diff-6c015c1d07296dbab7d1a69c15130511) and then [added back](https://github.com/hashicorp/packer/commit/1e853f9f1fcf36c54114db39039a50df7d4271c2#diff-6c015c1d07296dbab7d1a69c15130511) before.
The problem is that "url/net" is based on [rfc 1738](https://www.ietf.org/rfc/rfc1738.txt) and as such it always requires a host which is absent for local files. It also treats colons as a port separator, and this conflicts with Windows disk drive convention (e.g. C:\). In addition it does not support windows style separators (backslash).

[This line](https://github.com/hashicorp/packer/blob/master/common/config.go#L55) attempts to correct for the column after disk drive in windows path. Note the three slashes, the third one denotes the absense of host.

[This line](https://github.com/hashicorp/packer/blob/master/common/config.go#L73) acconts for paths like "\my\cool\path" that is windows paths that do not have disk drive in them.

Finally [this line](https://github.com/hashicorp/packer/blob/master/common/config.go#L80) I beleive was written in a rush. I examined git history for how it came to be, and I beleive that the reasoning for putting it in many years ago was wrong. I do not beleive that "net/url turns "C:/" into "/C:/"". This line was delivered in [this commit](https://github.com/hashicorp/packer/commit/7191c1f250f838ff65c786e36bfa66a92735698c#diff-6c015c1d07296dbab7d1a69c15130511). It was after:

- "file:///" prefix was being added for the "C:\something" path
- issue #5927 (see above) workaround was put in place
- issue #5927 was actually [fixed](https://github.com/golang/go/commit/39679ca88f8f37f6797daa9e866f8ba65ee18c9e#diff-8a328a849e3683316ce6ecce13b3fc73)

I do not beleive that it's "net/url" that turns "C:/" into "/C:/" as such. If you try to parse "C:\" without the "file:///" prefix it won't parse at all. I also beleive that whatever slash that this code removes is added back a few lines later by the workaround for #5927. Take a bit of time and compare these two pieces of code, the first one removes the slash and then the second one adds it back.

Now I'd like to mention that the actual fix for #5927 is:

- in url.String, if non-empty Path does not start with "/" prefix it with "/"
- but only if there is no schema, no host and no user. 

Since our code adds the "file" schema if there is no schema before the workaround runs, this fix is totally irrelevant for us.

The upshot of all this is that all those special case processing code used to result on windows in string urls returned by the `DownloadableURL` function with three slashes: "file:///C:/something...", but after the recent changes they are returned with just two slashes: "file://C:/something". In my opinion they are both incorrect, but the latter one is more incorrect since host part is missing. 

In this latter case `download.Get` function [here](https://github.com/hashicorp/packer/blob/master/common/download.go#L107) parses the drive name into the host part of the url, which is wrong and for the file schema it is naturally discarded. So when the downloader tries to download from this url the drive information is missing, which results in the code appearing to work if we are on the same drive, but fail if we are on a different drive.

**Quick Fix**
Put back the following lines removed in a [prior commit](https://github.com/hashicorp/packer/commit/4fb8a27879421803d82e2a5d7d437560a3208d74#diff-6a7d3eea6aa1a23b4ea371bdc85b36eb):


```
// Remove forward slash on absolute Windows file URLs before processing
if runtime.GOOS == "windows" && len(finalPath) > 0 && finalPath[0] == '/' {
  finalPath = finalPath[1:]
}
 ```

Comment out the following lines [from this file](https://github.com/hashicorp/packer/blob/v1.1.3/common/config.go#L78):


```
if len(url.Path) > 0 && url.Path[0] == '/' {
  url.Path = url.Path[1:len(url.Path)]
}
```

**Disclaimer**
I have never written a single line of golang code before today, I apologies, if there are any blunders above.

**Comment by arizvisa** [comment]
Btw, PR #2906 was written to deal with this sort of cross-platform weirdness with windows vs posix paths and removes all of those `runtime.GOOS == "windows"` hacks that are spotted throughout common/download.go. 

To deal with net/url not being designed to parse local file paths, PR #2906 uses the uri protocol to determine which path type the user has specified and then converts it to a local file path for that platform. This is handled by the LocalDownloader interface which is implemented by protocols that can be converted to a local file path.

Go 1.6 is when net/url started complaining about '\\' being illegal. Support for UNC paths in windows (which has paths prefixed with '\\\\') was fixed in the PR #2906 by implementing it as a separate smb:// protocol so that way net/url wouldn't prevent the user from specifying a UNC share as a url.

**Comment by AndrewSav** [comment]
@arizvisa Thank you for sharing. UNC paths is something that did not even occur to me.

So `DownloadableURL` needs to process:

- nix paths (/some/path)
- windows paths (c:\some\path)
- non-rooted paths (some/path or some\path)
- unc paths (\\server\some\path)
- http and https, and possibly ftp (http://some/path)

Note that the function accepts and returns string, but use net/url inside, then the downloader again uses net/url to parse it.

May be a custom abstraction for download path will work better?

By better I mean that issues are easier to find that they currently are. While `DownloadableURL`  has comments and is not very long, it is not very easy to read and understand which `if` is for which exactly combination of use cases. With custom abstraction it will also be easier to see which of the cases in the list above is being processed and why. - Just thinking out loud.

**Comment by AndrewSav** [comment]
@arizvisa - and the PR - looks awesome, I admire that after two years you are still hopeful ;)

**Comment by arizvisa** [comment]
Awesome that list of paths that you suggested are the exact ones that I was thinking about too when I first needed this thing to work consistently from Windows. (Tests for all of those are here : https://github.com/arizvisa/packer/blob/GH-2377/common/download_test.go#L439)

Hah, thx.. And lol, yeah...maybe not as hopeful (or young even) anymore. But at the very least, my attempt to participate in this community a few years ago has forced me to learn Go which ended up having more uses than I originally expected it to.

**Comment by arizvisa** [comment]
@AndrewSav, if you need a release that has these fixes/protocol that you need before they get around to merging these things upstream, I maintain releases and builds at https://github.com/arizvisa/packer/tree/release-$VERSION.

They include another set of patches that gives you extra options versus VMware such as control of the network and other misc things. My target audience is infosec people whom need to attach a kernel debugger, or need to ensure a template does not connect to the internet for any reason.

v1.1.3 should be there in a second, but to build it from src (since you're not a golang guy) you can just run `gox -os={windows,linux} -arch={i386,amd64} -output packer`

**Comment by AndrewSav** [comment]
@arizvisa well, I of course had to build the thing to be able to do the research above, no way around that, so I think I'll manage. Pointing out `gox` is also appreciated, I think I stumbled across it some months ago, but completely forgot about it by now.

**Comment by kaluznyt** [comment]
I've just installed packer v 1.1.3 (with choco) and wanted to build a windows image based on https://github.com/joefitzgerald/packer-windows/blob/master/windows_2012_r2.json , however I'm getting 

` virtualbox-iso: Downloading or copying: file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
    virtualbox-iso: Error downloading: CreateFile /Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso: The system cannot find the path specified.
==> virtualbox-iso: Guest additions download failed.
Build 'virtualbox-iso' errored: Guest additions download failed.`

Initially I thought that the problem is caused that I have a files on e: drive, but I moved to c: and get the same issue. 

I believe this thread is addressing this problem the fix is coming here https://github.com/hashicorp/packer/pull/5761 ? 

Any quick fixes/workarounds ?

Thanks

**Comment by AndrewSav** [comment]
Are the guest additions on C: drive?

**Comment by kaluznyt** [comment]
yes, guest additions are exactly under the path from the message that is 

file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso

![2018-01-05_09-06-53](https://user-images.githubusercontent.com/572605/34602499-d5d93610-f1f7-11e7-824f-d8fe76e50712.png)

**Comment by kaluznyt** [comment]
sorry, I was still running the command on e:, now it's working, ignore my comments

**Comment by kaluznyt** [comment]
So is it an issue or by design that iso file cannot be on a different drive than a script runs/template is placed ? I found this issue to happen when I run the script from command line on c: while json templates are on c as well, but url to iso is file://e:/... or e:/   but the same for c:/ works.

**Comment by SwampDragons** [comment]
@kaluznyt It's a mistake and we're going to fix it. :(

**Comment by kaluznyt** [comment]
thanks! I would really love to fix it but unfortunately I don't know GO (and packer codebase) at all so probably I wouldn't be too efficient in doing so... of course it's just nice to have, it still works! 👍

**Comment by AndrewSav** [comment]
@kaluznyt, this is going to be fixed by Hashicorp as SwampDragons indicated earlier.
I'd just like to give you some perspective regarding your "nice-to-have" statement.

Imagine a controlled environment, where your build machines (that is where you run packer for your image build automation) are built according to company standards.  These standards prescribe, that Operating System goes onto disc C: and everything else goes on another disk. This is not arbitrary: the machine is hosted is under a hypervisor and there are different tiers of data storage to provision disks upon, some are more expensive (fast) and some less. It is advisable to have OS on the fast disks, but it is not possible have all the images built on this machine have on the same fast disk. As a matter of compliance a regular check could be performed on disk C: so it does not has anything that is was not explicitly  provisioned for. As such, this setup does not give you an option to run packer from disk C:. And you suddenly go from "nice-to-have" to "I cannot use the latest version of packer until this is fixed".

So I'm glad, that it's nice-to-have *for you*, I'm just giving you a wider context.

**Comment by kaluznyt** [comment]
@AndrewSav I agree, it all depends on your current context. I'm currently more playing with it and learning (so it's 'nice to have' as I can easily workaround the issue, hence not critical for me at this point). But as you mentioned, if it's meant for production/enterprise use it should definitely handle such configurations.

**Comment by rgl** [comment]
Just to let you known that this was working fine in packer 1.1.1:

```
==> windows-2016-amd64-virtualbox: Downloading or copying Guest additions
    windows-2016-amd64-virtualbox: Downloading or copying: file:///C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
```

vs the broken 1.1.2:

```
==> windows-2016-amd64-virtualbox: Downloading or copying Guest additions
    windows-2016-amd64-virtualbox: Downloading or copying: file://C:/Program%20Files/Oracle/VirtualBox/VBoxGuestAdditions.iso
    windows-2016-amd64-virtualbox: Error downloading: CreateFile Program Files/Oracle/VirtualBox/VBoxGuestAdditions.iso: The system cannot find the path specified.
```
